Add ConcurrentArray class

saveriomiroddi opened the pull request:
Closes #482.

saveriomiroddi commented:
@st0012 I've tried a few design approaches, an I have a few questions:

1. I've tried to use sendMethod(), but it fails (see [code](https://github.com/saveriomiroddi/goby-dev/blob/3adfc08e501958d7c107f1921e9028331f42a3a5/vm/concurrent_array.go#L156)) - am I using it wrong? The args should be already in the stack, and I assume that directly passing blockFrame and sourceLine is correct?

2. with the current implementation (it's just an experiment) everything works, but the tests fail the CFP count - they return 1/2 instead of the expected 3/4 (go test ./vm -run ConcurrentArr); what's wrong?

3. finally, I can implement the class by using a pattern similar to what you've originally proposed for the concurrent hash, that is, by creating ConcurrentArrayObject as sublass of ArrayObject, and implementing a method like `lock()`, empty in ArrayObject, and with locking code in the ConcurrentArrayObject. This would require all the methods in ArrayObject to be converted to call that code. Is this your preferred implementation?

st0012 commented:
@saveriomiroddi 
1. I just pushed a fix for the first issue, can you help me check it?
2. Can you tell me what test functions failed, and the test case number? I guess it's because you call the builtin method's body, which is a Go function directly. So it doesn't create and push a new call frame properly. Using `sendMethod` should fix this issue.
3. Sorry I'm not sure I understand the third question, can you explain more or show the code?

BTW I think we can just make `ArrayObject` to have `sync.RWMutex` attribute by default, but don't use it. And then in `Concurrent::Array` we don't need to have another struct for concurrent array, I think this can reduce some work. However, I'm not sure if it will increase the memory usage of array objects significantly.

saveriomiroddi commented:
> I just pushed a fix for the first issue, can you help me check it?

The fix didn't work, or I'm using the wrong parameters; it's hard to know, because sendMethod() is undocumented. I'm following the example of the only occurrence in the codebase (inside `class.go`).

In order to test, just run `go ./vm -test ConcurrentArray`.

saveriomiroddi commented:
> Sorry I'm not sure I understand the third question, can you explain more or show the code?

Regarding this point, I will create a separate PR with an example.

st0012 commented on vm/concurrent_array.go:
```diff
@@ -0,0 +1,179 @@
+
+				// for testing sendMethod(), uncomment this block, and comment the following one.
+
+				// t.sendMethod(methodName, len(args) - 1, blockFrame, sourceLine)
```
You don't need to minus the arg count here. The minus one need to be added in `send`  because the first arg is method name, not a real argument for the method call.

st0012 replied:
I'll document that function since it seems that we will use it frequently in the future.

saveriomiroddi commented:
@st0012 is there any technical reason for you not testing the branch as suggested? I've applied this suggestion and it still doesn't work, so I'm blocked again by `sendMethod()`.

st0012 commented:
@saveriomiroddi Sorry, I'm a little sick and busy preparing my trip. I'll try it ASAP.

st0012 commented:
@saveriomiroddi After a quick try, I just thought a possible reason: Assume we call `sendMethod` inside `[]`, we'll get into `Concurrent::Array#[]`. And it'll still call the same method `[]` again just like an infinite loop. But the first call already consumed the argument, so the second call won't have any and causes index panic.

saveriomiroddi commented:
@st0012 ok, thanks! :+1:

st0012 commented:
@saveriomiroddi I found that the CFPs you got are actually correct, they don't necessarily equal to the error line number. In fact, the CFP, SP and error line number basically won't effect each other.

The CFP stands for `call frame pointer`, it's like stack pointer but for call frame stack. Goby creates a call frame for each method call and push it to the call frame stack. Once the method call is finished it'd be popped from call frame stack. So we can assume that if every program is executed successfully the CFP will be 0 because every call frame should be executed and popped.

That's see an example:

```ruby
def foo
  10
end

foo
```

1. Before the `foo` is being called, the call frame stack would be empty and the CFP would be 0.
2. After the method `foo` is called its call frame will be pushed on the stack, and the CFP will increase by 1.
3. Once the `foo` method is finished, its frame will be popped. So the call frame stack is now empty and the CFP will decrease by 1 and back to 0. (This is why every method test expect CFP to be 0.)

However, if any method failed, its call frame will stay on the stack and the CFP will stay the same as well. This is why most of the fail test case's CFP is 1, because that method failed and its frame is not popped.

(To be continued)

saveriomiroddi commented:
> I found that the CFPs you got are actually correct, they don't necessarily equal to the error line number. In fact, the CFP, SP and error line number basically won't effect each other.

@st0012 thanks for the information :smile: I will update all the CFPs in the test suite.

saveriomiroddi commented:
@st0012 I actually dug a bit more, and the CFP expectations with value of 2 are actually bugs in the code. I'll commit the fixes and show you.

saveriomiroddi commented:
@st0012 Ready for review and then merge. I've corrected the block frame bugs in the Array class, and updated the UTs, so now the CFP make sense (they all should be `1`).

saveriomiroddi commented:
Found at least another bug in Array, will fix.

saveriomiroddi commented:
ok. fixed that bug as well. everything's ready.

st0012 commented:
@saveriomiroddi I actually keep the block frame on purpose so that we can know if the error is happened in the method or in the method's block. It doesn't make any difference for now, but I'll need this information for #474

saveriomiroddi commented:
@st0012 thanks, I've restored the previous structure and CTF expectations.

st0012 commented:
@saveriomiroddi The TravisCI failed, I think you should rebuild it

saveriomiroddi commented:
@st0012 hmm, the build didn't start all - it only reported [an error](https://s3.amazonaws.com/archive.travis-ci.org/jobs/295201796/log.txt?X-Amz-Expires=30&X-Amz-Date=20171031T110143Z&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAJRYRXRSVGNKPKO5A/20171031/us-east-1/s3/aws4_request&X-Amz-SignedHeaders=host&X-Amz-Signature=510b7be6b6ebf1ad9d3b78c57f0e3e0d073f40b12ecaefbc4182ed118c4ca6da).

I've restared it - I'll check later the outcome.
